core/src/layers/throttle.rs

```diff
@@ -0,0 +1,332 @@
+    }
+
+    async fn append(&self, path: &str, args: OpAppend) -> Result<(RpAppend, Self::Appender)> {
+        let limiter = self.rate_limiter.clone();
+
+        self.inner
+            .append(path, args)
+            .await
+            .map(|(rp, a)| (rp, ThrottleWrapper::new(a, limiter)))
+    }
+
+    async fn list(&self, path: &str, args: OpList) -> Result<(RpList, Self::Pager)> {
+        self.inner.list(path, args).await
+    }
+
+    fn blocking_read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::BlockingReader)> {
+        let limiter = self.rate_limiter.clone();
+
+        self.inner
+            .blocking_read(path, args)
+            .map(|(rp, r)| (rp, ThrottleWrapper::new(r, limiter)))
+    }
+
+    fn blocking_write(&self, path: &str, args: OpWrite) -> Result<(RpWrite, Self::BlockingWriter)> {
+        let limiter = self.rate_limiter.clone();
+
+        self.inner
+            .blocking_write(path, args)
+            .map(|(rp, w)| (rp, ThrottleWrapper::new(w, limiter)))
+    }
+
+    fn blocking_list(&self, path: &str, args: OpList) -> Result<(RpList, Self::BlockingPager)> {
+        self.inner.blocking_list(path, args)
+    }
+}
+
+pub struct ThrottleWrapper<R> {
+    inner: R,
+    limiter: SharedRateLimiter,
+}
+
+impl<R> ThrottleWrapper<R> {
+    pub fn new(inner: R, rate_limiter: SharedRateLimiter) -> Self {
+        Self {
+            inner,
+            limiter: rate_limiter,
+        }
+    }
+}
+
+impl<R: oio::Read> oio::Read for ThrottleWrapper<R> {
+    fn poll_read(&mut self, cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<Result<usize>> {
+        // TODO: How can we handle buffer reads with a limiter?
+        self.inner.poll_read(cx, buf)
+    }
+
+    fn poll_seek(&mut self, cx: &mut Context<'_>, pos: SeekFrom) -> Poll<Result<u64>> {
+        self.inner.poll_seek(cx, pos)
+    }
+
+    fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Option<Result<Bytes>>> {
+        self.inner.poll_next(cx)
+    }
+}
+
+impl<R: oio::BlockingRead> oio::BlockingRead for ThrottleWrapper<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        // TODO: How can we handle buffer reads with a limiter?
+        self.inner.read(buf)
+    }
+
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
+        self.inner.seek(pos)
+    }
+
+    fn next(&mut self) -> Option<Result<Bytes>> {
+        self.inner.next()
+    }
+}
+
+#[async_trait]
+impl<R: oio::Write> oio::Write for ThrottleWrapper<R> {
+    async fn write(&mut self, mut bs: Bytes) -> Result<()> {
+        // Note: for self.inner.write(bs), it's possible that the given bs length is less than the total content length.
+        // Users will call write multiple times to write the whole data.
+
+        let mut buf_length = NonZeroU32::new(bs.len() as u32).unwrap();
+
+        while !bs.is_empty() {
+            match self.limiter.check_n(buf_length) {
+                Ok(_) => self.inner.write(bs.split_to(buf_length as usize)).await,
```

Comment:
It's not a good idea for me to split write operations because doing so changes the semantics of the `write` operation. It's required that `write` been called only once.

For object storage services like `s3`, calling `write` could change underlying request from `PutObject` to `MultipartUpload` which is not acceptable.

Reply:
That's what I'm afraid of, as I can't come up with a solution without splitting operations right now :(

Reply:
We can sleep until we get enough cells.

Reply:
This is quite a dilemma. Cause if the length of Bytes is larger than the burst size, then the bytes will never be passed through.
After looking up a Quota example [here](https://docs.rs/governor/latest/governor/struct.Quota.html#examples), it seems that the only solution for this is to set the `replenish_rate` and `burst_size` separately, but it's ugly: 
**The burst size should be larger than any possible byte length to allow it to pass through**. The next operation will have to wait until a sufficient cells is replenished based on the `replenish_rate`. Although it is limited to a given bandwidth over a wider time period, the flow of bytes is **not smooth** if the user operates on very large data. 
Let me give an example here:
```rust
// User wants to limit bandwidth to 10 KiB/s, but their file size is close to 10 MB.
let q = Quota::per_second(nonzero!(10*1024u32)).allow_burst(nonzero!(10000*1024u32))
// replenish rate close to 0.1ms per byte
q.replenish_interval()  // 97.656µs
// The entire maximum burst size will be restored if no cells are let through for 1000 seconds.
q.burst_size_replenished_in()  // 999.99744s

// That means if the user has multiple operations on files of size ~10MB, 
// each operation will have to wait ~1000 seconds before executing the next operation.
```

Reply:
Based our current API, it's the only way we can do this.

Reply:
Change throttle strategy based on your conclusion.